### DISCUSSION TITLE
How do I list target dependencies in quarto.yml? [help]

### DISCUSSION THREAD

**Discussion by daranzolin** [ropensci/targets#1255] [Help] [answered]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

In [the documentation I read:](https://books.ropensci.org/targets/literate-programming.html)

> tar_quarto() looks for target dependencies in all the source documents (e.g. listed in _quarto.yml)...

How do I specify the dependencies in `_quarto.yml`? What is the 'key' for the key-value pair? I currently have:

```
 project:
  post-render:
    - copy_report.R
```

**Comment by joelnitta** [comment]
I am honestly not quite sure what the reference to `_quarto.yml` is about (perhaps parameterized reports?) but generally what I do to load targets built during the workflow into a quarto report is include a chunk with calls to `tar_load()` and `tar_read()`, for example like [this](https://github.com/joelnitta/penguins-targets/blob/main/penguin_report.qmd#L10-L17):

```{r}
#| label: load
#| message: false
targets::tar_load(model_predictions)
targets::tar_load(model_summaries)
```

**Reply by daranzolin** [reply]
I see that's the default way of loading targets, but my qmds are not in the project root, but a `Reports` directory one level 'deeper.' 

So I load my targets like this:

```r
proj_read <- purrr::partial(targets::tar_read, store = path(here(), "_targets"))
model_predictions <- proj_read(model_predictions)
model_summaries <- proj_read(model_summaries)
```

And so in `_targets.R` I have:

```r
tar_quarto(my_report, "Reports/my_report.qmd", execute_params = list(...))
```

But this is leading to some awkwardness and I've also ended up wrapping `quarto::quarto_render` in some other functions. I thought maybe if specifying the targets in `_quarto.yml` would be simpler and more elegant.

**Comment by joelnitta** [comment]
I'm not sure, but I suspect calling `tar_read()` like that within `purrr::partial()` could make it tough for targets to detect dependencies.

I've also had problems with quarto documents in nested folders, in particular when trying to get the output rendered to a subfolder.

I tried various things, but found that in the end it was easier to just stick the `qmd` in the project root and live with it.

**Reply by daranzolin** [reply]
I was just about to reach the same conclusion! This was my last attempt to preserve my original project structure. Thanks for responding!

**Comment by gorkang** [comment, marked as answer]
Not sure if this is what you need, but I have qmd reports in subfolders that are fed from other targets:


My _quarto.yml:

```
project:
  output-dir: outputs
  render:
    - Qmd/REPORT.qmd
```

_targets.R
```
  tarchetypes::tar_quarto(NAME_target,
                          path = getwd()
   )
```

And in a chunk in Qmd/REPORT.qmd

```
targets::tar_load(DF_raw, store = here::here("_targets/"))
```

**Reply by daranzolin** [reply]
Does `tarchetypes::tar_knitr_deps("Qmd/REPORT.qmd")` return the target dependencies?

**Reply by joelnitta** [reply]
Cool! This works for me, and I do get the dependency that I loaded with `tar_load()` in my qmd when I run `tarchetypes::tar_knitr_deps()`.

**Reply by joelnitta** [reply]
(note that `_quarto.yml` goes in the project root, and that I don't think you actually need `path = getwd()` for  `tarchetypes::tar_quarto()`, since it uses the project directory for `path` by default anyways).

**Reply by daranzolin** [reply]
This works for me too! I think this is the way to go, I was close with my purrr::partial, but this makes much more sense. Thanks @gorkang !

**Comment by Aariq** [comment]
I've had some success with putting `tar_config_set(store = "../_targets/")` at the top of a .qmd not in the root folder, and then just using `tar_read()` like "normal".  Of course, filepaths from targets with `format = "file"` won't be relative to the .qmd document, so then I do things like `include_graphics(tar_read(my_figure) |> fs::path_rel("docs"))`.  I'm still not sure what the "correct" way to do all this is though ('all this' meaning rendering a quarto/rmd doc not in the project root with a targets pipeline while still having the report work well when interactively running code chunks)

**Reply by Aariq** [reply]
Oh, nope, nevermind.  Just realized that doesn't work because it creates a second _targets store when you run the pipeline.  🤦‍♂️.  I've also tried using _quarto.yml to set the execute dir to be "project", but I can't remember why that didn't work.  Would be nice to have a vignette of how to do this the "right" way.

**Reply by daranzolin** [reply]
Cosign! Not a huge deal, but it seems reasonable to want your qmd reports in separate directory and write:

```r
tar_quarto(my_report, "Reports/my_report.qmd")
```

in `_targets.R`.
